internal/testing/cmdtest/cmdtest.go

```diff
@@ -0,0 +1,610 @@
+// See the TestFile documentation for the syntax.
+//
+// To test, first read the file:
+//
+//    tf, err := cmdtest.Read("test.ct")
+//
+// Then configure the resulting TestFile by adding commands or enabling
+// debugging features. Lastly, call TestFile.Compare:
+//
+//    diff := tf.Compare()
+//
+// The string returned by Compare will be non-empty if there were any
+// errors or differences in output.
+//
+// After confirming that the output differences are intentional, you can
+// update the original file by calling TestFile.Update:
+//
+//    err := tf.Update()
+package cmdtest
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// A TestFile describes a single file that may contain multiple test cases.
+// File format:
+//
+// Before the first line starting with a '$', empty lines and lines beginning with
+// "#" are ignored.
+//
+// A sequence of consecutive lines starting with '$' begin a test case. These lines
+// are commands to execute. See below for the valid commands.
+//
+// Lines following the '$' lines are command output (merged stdout and stderr).
+// Output is always treated literally. After the command output there should be a
+// blank line. Between that blank line and the next '$' line, empty lines and lines
+// beginning with '#' are ignored. (Because of these rules, cmdtest cannot
+// distinguish trailing blank lines in the output.)
+//
+// Syntax of a line beginning with '$':
+// A sequence of space-separated words (no quoting is supported). The first word is
+// the command, the rest are its args. If the next-to-last word is '<', the last word
+// is interpreted as a file and becomes the standard input to the command. This input
+// redirection is only supported for commands run with exec.Command, not those in the
+// Commands map.
+//
+// By default, commands are expected to succeed, and the test will fail
+// otherwise. However, commands that are expected to fail can be marked
+// with a " --> FAIL" suffix.
+//
+// The cases of a test file are executed in order, starting in a freshly
+// created temporary directory.
+//
+// The built-in commands (initial contents of the Commands map) are:
+//
+//   cd DIR
+//   cat FILE
+//   mkdir DIR
+//   setenv VAR VALUE
+//   echo ARG1 ARG2 ...
+//   echof FILE ARG1 ARG2 ...
+//
+// These all have their usual Unix shell meaning, except for echof, which writes its
+// arguments to a file (output redirection is not supported). All file and directory
+// arguments must refer to the current directory; that is, they cannot contain
+// slashes.
+//
+// If the first word of a command line is not one of the above built-ins, or any
+// additional commands added to the TestFile.Commands map before execution, then
+// cmdtest runs the command using the exec package. Recall that the exec package
```

Comment:
Do we have a use case for this (exec'ing arbitrary executables)?

I started out with this in the `gocdk` tests but ended up removing it because I didn't need it, and because it's difficult to make portable.

Reply:
Each of the sample CLIs is an executable, and I can make them available just by setting PATH. However, I could also make it easy to install a custom command that invokes that particular binary.

I tested on windows (by pushing a commit that hacked .travis.yml to run windows on PRs) and I _thought_ it just worked, but maybe I messed up. Do you have to change the binary name to end in ".exe" or does the exec package handle that for you? What else is non-portable?

Reply:
Yes, you have to change the binary name to end in `.exe`.

If you're using this primarily as a way to run the executable under test, I'd suggest requiring the test to install it in the `Commands` map; maybe a helper for doing so would make it easy. The non-portable thing you've exposed here is that I can easily write tests that call non-portable things like `ls`.

Reply:
(and then remove the ability to `exec` arbitrary executables; everything should have to be built-in).

Reply:
> Yes, you have to change the binary name to end in `.exe`.

You don't have to do this. I logged into a Windows GCP machine and ran the following program (and I deserve a medal for it):
```
package main

import "fmt"
import "os/exec"

func main() {
	out, err := exec.Command("go", "version").CombinedOutput()
	fmt.Println(err)
	if err == nil {
		fmt.Println(string(out))
	}
}
```
Notice that it doesn't say `go.exe`. It still worked. I also tried the absolute path `c:\go\bin\go` without the `.exe`. That also worked.